gpt_engineer/ai.py

```diff
@@ -23,72 +39,134 @@ class TokenUsage:
-        self.model = model
+        if model_dir:
+            self.model_dir = Path(model_dir)
+        else:
+            # __file__ is "<package_dir>/gpt-engineer/ai.py"
+            # therefore .parent.parent/models is "<package_dir>/models"
+            self.model_dir = Path(__file__).resolve().parent.parent / "models"
+        self.model_name = fallback_model(model_name)
+        self.llm = None
+
+        llm_filename = self.model_dir / f"{model_name}.yaml"
+        try:
+            logging.info(f"LLM file name: {llm_filename}")
+            self.llm = load_llm(llm_filename)
+        except Exception as e:
+            raise RuntimeError(
+                f"Unable to load LLM {model_name} from file {llm_filename}", e
+            )
 
         # initialize token usage log
         self.cumulative_prompt_tokens = 0
         self.cumulative_completion_tokens = 0
         self.cumulative_total_tokens = 0
         self.token_usage_log = []
 
+        # ToDo: Adapt to arbitrary model, currently assumes model using tiktoken
         try:
-            self.tokenizer = tiktoken.encoding_for_model(model)
+            self.tokenizer = tiktoken.encoding_for_model(model_name)
         except KeyError:
             logger.debug(
-                f"Tiktoken encoder for model {model} not found. Using "
+                f"Tiktoken encoder for model {model_name} not found. Using "
                 "cl100k_base encoder instead. The results may therefore be "
                 "inaccurate and should only be used as estimate."
             )
             self.tokenizer = tiktoken.get_encoding("cl100k_base")
 
-    def start(self, system, user, step_name):
+    def start(self, system: str, user: str, step_name: str) -> List[Message]:
         messages = [
-            {"role": "system", "content": system},
-            {"role": "user", "content": user},
+            SystemMessage(content=system),
+            HumanMessage(content=user),
         ]
-
         return self.next(messages, step_name=step_name)
 
-    def fsystem(self, msg):
-        return {"role": "system", "content": msg}
+    def fsystem(self, msg: str) -> SystemMessage:
+        return SystemMessage(content=msg)
 
-    def fuser(self, msg):
-        return {"role": "user", "content": msg}
+    def fuser(self, msg: str) -> HumanMessage:
+        return HumanMessage(content=msg)
 
-    def fassistant(self, msg):
-        return {"role": "assistant", "content": msg}
+    def fassistant(self, msg: str) -> AIMessage:
+        return AIMessage(content=msg)
 
-    def next(self, messages: List[Dict[str, str]], prompt=None, *, step_name=None):
+    def combine_messages(self, messages: List[Message]) -> str:
+        msg_dict = messages_to_dict(messages)
+        logging.debug(msg_dict)
+        prompt = "\n".join(
+            "%s: %s" % (md["type"], md["data"]["content"]) for md in msg_dict
+        )
+        logging.debug("Prompt: " + prompt)
+        return prompt
```

Comment:
I'd prefer we get rid of this and only support GPT4 for now. It's by far the best code generation model.

Reply:
Agreed & changed